src/visualization.c

```diff
@@ -46,17 +67,60 @@ void __cdecl SAAddPCMData(void *PCMData, int nch, int bps, int timestamp) {
 
   const uint16_t *const interleaved = (uint16_t *)PCMData;
 
+  // For waveform: De-interleave and scale to 8bit
   for (int i = 0; i < VIS_FRAMES; i++) {
-    // De-interleave and scale to 8bit
     g_active_vis_module->waveformData[0][i] = interleaved[2 * i] / 256;
     g_active_vis_module->waveformData[1][i] = interleaved[2 * i + 1] / 256;
+  }
 
-    // TODO this is debugging data, do spectral analysis instead
-    g_active_vis_module->spectrumData[0][i] = i * (256 - 1) / (VIS_FRAMES - 1);
-    g_active_vis_module->spectrumData[1][i] =
-        (VIS_FRAMES - i - 1) * (256 - 1) / (VIS_FRAMES - 1);
+  // For spectrum: De-interleave, do spectral analysis, and scale to 8bit
+  for (int channel = 0; channel < 2; channel++) {
+    kiss_fft_scalar scalar_in[VIS_FRAMES * 2];
+    kiss_fft_cpx cx_out[VIS_FRAMES * 2];
+    kiss_fft_scalar *const scalar_in_first_half = scalar_in;
+    kiss_fft_scalar *const scalar_in_second_half = scalar_in + VIS_FRAMES;
+
+    // Prepare FFT input
+    for (int i = 0; i < VIS_FRAMES; i++) {
+      // De-interleave and apply Hann window function
+      scalar_in_first_half[i] =
+          (kiss_fft_scalar)g_prev_interleaved[2 * i + channel] *
+          hann_factor(i, VIS_FRAMES * 2);
+      scalar_in_second_half[i] =
+          (kiss_fft_scalar)(int16_t)interleaved[2 * i + channel] *
+          hann_factor(i + VIS_FRAMES, VIS_FRAMES * 2);
+    }
+
+    // Apply FFT
+    kiss_fftr(g_kiss_fft_cfg, scalar_in, cx_out);
+
+    // Post-process FFT output, in particular do scaling:
+    // - We need to compensate the scaling that FFT did:
+    //   factor "1.0f / (VIS_FRAMES / 2)".
+    // - We need to convert range from 0..2^15-1 to 0..2^8-1:
+    //   factor "1.0f / INT16_MAX * UINT8_MAX".
+    // - The rest is compensation of the Hann window plus additional zoom:
+    //   factor "5.0f".
+    const kiss_fft_scalar amplitude_scale =
```

Comment:
@hartwork It's not really important for perf here, but you could make it `static` since no variable are used here. Thus the value is computed only once at the launch of the program.

Reply:
@BenPyton I was counting on the compiler to fully resolve this constant here, and it turns out it does do that for this very case in practice ([proof](https://gcc.godbolt.org/#g:!((g:!((g:!((h:codeEditor,i:(filename:'1',fontScale:14,fontUsePx:'0',j:1,lang:___c,selection:(endColumn:2,endLineNumber:12,positionColumn:2,positionLineNumber:12,selectionStartColumn:2,selectionStartLineNumber:12,startColumn:2,startLineNumber:12),source:'%23define+VIS_FRAMES+576%0A%0A%23include+%3Cstdint.h%3E%0A%0Aint+main(int+argc,+char+**argv)+%7B%0A++const+float+one+%3D+111.111f%3B%0A++const+float+two+%3D+222.222f%3B%0A++const+float+amplitude_scale+%3D+1.0f+/+(VIS_FRAMES+/+2)+*+5.0f+/+INT16_MAX+*+UINT8_MAX%3B%0A++printf(%22Hello+%25f+%25d+%25f%5Cn%22,+one,+two,+amplitude_scale)%3B%0A%0A++return+0%3B%0A%7D%0A'),l:'5',n:'0',o:'C+source+%231',t:'0')),k:50,l:'4',n:'0',o:'',s:0,t:'0'),(g:!((h:compiler,i:(compiler:cg122,deviceViewOpen:'1',filters:(b:'0',binary:'1',binaryObject:'1',commentOnly:'0',demangle:'0',directives:'0',execute:'1',intel:'1',libraryCode:'0',trim:'1'),flagsViewOpen:'1',fontScale:14,fontUsePx:'0',j:1,lang:___c,libs:!(),options:'',selection:(endColumn:1,endLineNumber:1,positionColumn:1,positionLineNumber:1,selectionStartColumn:1,selectionStartLineNumber:1,startColumn:1,startLineNumber:1),source:1),l:'5',n:'0',o:'+x86-64+gcc+12.2+(Editor+%231)',t:'0')),k:50,l:'4',n:'0',o:'',s:0,t:'0')),l:'2',n:'0',o:'',t:'0')),version:4)). So this variable does not exist at runtime any more.